### PR TITLE
Migrate to Jest for JS unit testing

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

While discussing #428 and the exceptions made for QUnit, I thought I'd try a short at migrating these tests over to Jest.

So far it works really well. I only had to skip a few tests because of various reasons:

a) Node is missing some locale data
b) Some tests were inconsistent
c) Tests have weird dependencies on global variables

Addresses #524 and #782.

## Relevant technical choices

* Sets up new Jest config for unit tests
* Removes QUnit dependency and tests
* Migrates  all tests over to new Jest setup
* Uses proper imports instead of accessing globals

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

**Comment by tofumatt** [comment]
@googlebot I consent.

**Comment by tofumatt** [comment]
Getting some bizarre/cryptic/unclear errors at this point in Travis, see: https://travis-ci.com/google/site-kit-wp/jobs/250737922#L1632

I'm going to clear the cache but locally I've got this passing and not failing on what Travis is failing on. I'm pretty confused here 😆

**Comment by tofumatt** [comment]
Okay! We've got this all working now with Jest, E2E tests, visual tests, etc.

I've checked through the ACs and this doesn't increase the bundle size or include test files, and things still work in all supported WP versions. This is now ready for review; sorry for the massive diff (though most of it is `package-lock.json` changes). It's also quite nice having the `@wordpress/` imports setup; thanks for that @swissspidy.

(I couldn't assign @swissspidy for review here, but it'd be great if you had a look at this and made sure it looks okay. It's been rebased on top of `develop` and I think it was fine, but it'd be good to know if I accidentally left anything out.)

Yay! 🎉

File: tests/backstop/engine_scripts/puppet/clickAndHoverHelper.js [hunk 1]
```diff
@@ -6,20 +6,23 @@ module.exports = async ( page, scenario ) => {
...
 	if ( hoverSelector ) {
+		// eslint-disable-next-line no-unused-vars
 		for ( const hoverSelectorIndex of [].concat( hoverSelector ) ) {
 			await page.waitFor( hoverSelectorIndex );
 			await page.hover( hoverSelectorIndex );
 		}
 	}
 
 	if ( clickSelector ) {
+		// eslint-disable-next-line no-unused-vars
```
Comment (aaemnnosttv) [inline tests/backstop/engine_scripts/puppet/clickAndHoverHelper.js]: Double check if these `eslint-disable-next-line`s are necessary

File: tests/js/setup-globals.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import React from 'react';
+
+if ( global.Intl ) {
+	// Check for locales other than American English.
+	if ( ! intlLocalesSupported( [ 'en-US', 'de-DE' ] ) ) {
+		global.Intl.NumberFormat = intl.NumberFormat;
+		global.Intl.DateTimeFormat = intl.DateTimeFormat;
+	}
+} else {
+	throw new Error( 'Your version of node is very old and does not support `Intl`. Please use at least node 10.' );
```
Comment (aaemnnosttv) [inline tests/js/setup-globals.js]: Is Node 10 required? It seems that `Intl` is mentioned as far back as [Node 8 in the docs ](https://nodejs.org/docs/latest-v8.x/api/intl.html#intl_internationalization_support) but seems to depend on how Node was built.

File: webpack.config.js [hunk 3]
```diff
@@ -71,9 +71,7 @@ const externals = {
 [
 	...externalPackages,
 ].forEach( ( name ) => {
-	externals[ `@wordpress/${ name }` ] = {
-		this: [ 'wp', camelCaseDash( name ) ],
-	};
+	externals[ `@wordpress/${ name }` ] = [ 'wp', camelCaseDash( name ) ];
```
Comment (aaemnnosttv) [inline webpack.config.js]: Nice 👍

File: assets/js/components/dashboard-splash/dashboard-splash-app.js [hunk 4]
```diff
@@ -22,6 +22,12 @@
 import DashboardSplashMain from 'GoogleComponents/dashboard-splash/dashboard-splash-main';
 import { Suspense as ReactSuspense, lazy as ReactLazy } from 'react';
 
+/**
+ * WordPress dependencies
+ */
+import { Component, Fragment, Suspense as WPSuspense, lazy as WPlazy } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/components/dashboard-splash/dashboard-splash-app.js]: I'm not sure that ever got merged; at any rate it was weird because even the version of React we load depends on what Gutenberg/other scripts expose. I'll have a look though. 👍

File: assets/js/components/data-table.js [hunk 5]
```diff
@@ -22,13 +22,16 @@
...
 import Link from 'GoogleComponents/link';
-
-const {
+import {
 	each,
 	debounce,
 	trim,
-} = lodash;
-const { Component, createRef } = wp.element;
+} from 'lodash';
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/components/data-table.js]: It uses the global as well.

File: tests/js/setup-globals.js (hunk 2 above)
Comment (tofumatt) [reply tests/js/setup-globals.js]: I think 10 is safer, but it's also the version we specify in our `.nvmrc` anyway, so telling users to use it I think is fine.

File: assets/js/util/index.js [hunk 6]
```diff
@@ -18,34 +18,35 @@
...
 	isUndefined,
 	unescape,
 	deburr,
 	toLower,
 	trim,
-} = lodash;
+	trimEnd,
+} from 'lodash';
+import data, { TYPE_CORE } from 'GoogleComponents/data';
+import SvgIcon from 'GoogleUtil/svg-icon';
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/util/index.js]: Yeah, that's _all_ over the codebase. I think it's a separate issue worth addressing, but it's the norm in the codebase right now.

[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/util/index.js]: 10-4. Is that a module resolution issue with the aliases that ESLint isn't picking up on? Can you open a new issue to address this?

[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/util/index.js]: Filed #780.

File: assets/js/util/index.js [hunk 7]
```diff
@@ -445,20 +447,6 @@ export const getSiteKitAdminURL = ( page, args ) => {
...
- */
-export const isFrontendIframeLoaded = () => {
-	const iframe = document.getElementById( 'sitekit_fe_load_check' );
-	if ( iframe ) {
-		return iframe;
-	}
-
-	return false;
-};
-
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/util/index.js]: I see it imported and exported but not in use anywhere. Seems it was used in https://github.com/google/site-kit-wp/blob/6d344ce0c83560df41f6f8992a73b40271be68c4/tests/qunit/assets/js/util/index.js but but I'm not _exactly_ clear what it's trying to do. It doesn't seem to have been migrated over, perhaps because Jest isn't usually rendering full pages/browser DOM so it would have been harder to test?

Though the existing tests don't even seem to be testing many states 🤷‍♂. Seems safe to me to remove, but let me know if I'm off-base here.

File: assets/js/util/test/toggleConfirmModuleSettings.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ * Internal dependencies
+ */
+import { toggleConfirmModuleSettings } from '../';
+
+describe( 'toggleConfirmModuleSettings', () => {
+	it( 'should toggle settings for analytics module', ( ) => {
+		const googlesitekit = {
+			modules: {
+				analytics: {
+					settings: { accountId: '12345678' },
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/util/test/toggleConfirmModuleSettings.js]: Indeed, it is a regression. Cheers.

File: assets/js/googlesitekit-module.js [hunk 9]
```diff
@@ -22,26 +22,25 @@
...
  */
-import ModuleApp from './components/module-app';
-
-const { setLocaleData } = wp.i18n;
-const { doAction, applyFilters } = wp.hooks;
-const { Component, render, Fragment } = wp.element;
-let { Suspense, lazy } = wp.element;
+import { setLocaleData } from '@wordpress/i18n';
+import { doAction, applyFilters } from '@wordpress/hooks';
+import { Component, render, Fragment, Suspense as WPSuspense, lazy as WPlazy } from '@wordpress/element';
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/googlesitekit-module.js]: Agreed, but I think it's being worked on outside and best to do separate from here at this point. We really need to establish minimum React/`@wordpress/*` versions and figure out how we load them to sort this out...

**Comment by tofumatt** [comment]
Removed `wp` from our globals in ESLint and thus the `wp` global usage now fails. Everything should be addressed and ready for another review. Cheers!

File: assets/js/util/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/util/index.js]: This is related to an older version of (existing) tag detection which used an iframe instead of `fetch`. This doesn't appear to be used anywhere like you said so it should be removed 👍

**Comment by tofumatt** [comment]
I'm seeing three failures setting up Analytics from `stories/settings.stories.js` but I'm unclear on why the JSON responses there would have changed. @adamsilverstein: any chance I'm missing something needed to load this test data properly? Sorry, but I haven't changed much in that file so I'm not clear what's up.

I wonder if we need to be better mocking `@wordpress/api-fetch` now that we're importing it rather than relying on the global? 🤷‍♂

**Comment by felixarntz** [comment]
@swissspidy Can you take a look at the additional changes?

**Comment by aaemnnosttv** [comment]
@tofumatt the problem is that it's making an API request using `apiFetch` which is mocked in the storybook config using the `wp` global (along with a bunch of other stuff).

image

Looks like the eslint rule for the `wp` global is being excluded here.

https://github.com/google/site-kit-wp/blob/102e534484f8b9735621c459eff72586cfde286b/.storybook/config.js#L86-L91

Can we mock the response of this module resolution for storybook only?

**Comment by felixarntz** [comment]
@tofumatt Note that there is a conflict with `develop`.

**Comment by tofumatt** [comment]
Conflict fixed; it was a new QUnit test. I ported it over to Jest so this should be set now!

File: .eslintrc.json [hunk 10]
```diff
@@ -21,8 +21,7 @@
     "lodash": "readonly",
     "googlesitekitCurrentModule": "readonly",
     "gtag": "readonly",
-    "process": "readonly",
-    "wp": "readonly"
```
Comment (felixarntz) [inline .eslintrc.json]: Should this not remain `readonly`?

File: assets/js/modules/pagespeed-insights/setup.js [hunk 11]
```diff
@@ -0,0 +1,198 @@
+/**
+ * TagmanagerSetup component.
```
Comment (felixarntz) [inline assets/js/modules/pagespeed-insights/setup.js]: This is old code. There shouldn't be a setup component for PSI.

File: assets/js/util/test/__snapshots__/showErrorNotification.js.snap [hunk 12]
```diff
@@ -0,0 +1,9 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (felixarntz) [inline assets/js/util/test/__snapshots__/showErrorNotification.js.snap]: Are these files needed under version control? I'm not sure what this is for.

File: assets/js/components/dashboard-splash/dashboard-splash-app.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixarntz) [reply assets/js/components/dashboard-splash/dashboard-splash-app.js]: This still needs to be revisited, it should import from `util/react-features`.

File: assets/js/components/dashboard-splash/dashboard-splash-app.js [hunk 13]
```diff
@@ -31,13 +37,11 @@ import { sendAnalyticsTrackingEvent } from 'GoogleUtil';
...
 
-const { Component, Fragment } = wp.element;
-let { Suspense, lazy } = wp.element;
-const { __ } = wp.i18n;
-
 // Check for `Suspense` and `lazy` in `wp.element`; versions before 2.4.0 did
 // not include either, so we need to fallback to the React versions. See:
 // https://github.com/WordPress/gutenberg/blob/master/packages/element/CHANGELOG.md#240-2019-05-21
+let Suspense = WPSuspense;
+let lazy = WPlazy;
```
Comment (felixarntz) [inline assets/js/components/dashboard-splash/dashboard-splash-app.js]: See above, this can be removed following the above change.

File: assets/js/util/test/getReAuthUrl.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+ * Internal dependencies
+ */
+import { getReAuthURL } from '../';
+
+const createSiteKit = ( apiKey ) => {
+	return {
+		admin: {
+			adminRoot: 'http://sitekit.withgoogle.com/wp-admin/admin.php',
+			apikey: apiKey,
+			connectUrl: 'http://sitekit.withgoogle.com/wp-admin/admin.php?googlesitekit_connect=1&nonce=12345&page=googlesitekit-splash',
```
Comment (felixarntz) [inline assets/js/util/test/getReAuthUrl.js]: This should be `connectURL`.

File: assets/js/util/test/getReAuthUrl.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
+/**
+ * Internal dependencies
+ */
+import { getReAuthURL } from '../';
+
+const createSiteKit = ( apiKey ) => {
```
Comment (felixarntz) [inline assets/js/util/test/getReAuthUrl.js]: A bit unrelated, but let's remove the `apiKey` references here entirely as they were only necessary when we actually used one.

File: assets/js/util/test/getReAuthUrl.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+			getReAuthURL( 'pagespeed-insights', false, googlesitekit )
+		).toStrictEqual(
+			'http://sitekit.withgoogle.com/wp-admin/admin.php?page=googlesitekit-dashboard&slug=pagespeed-insights&notification=authentication_success'
+		);
+
+		googlesitekit = createSiteKit( 'abc123' );
+
+		expect(
+			getReAuthURL( 'pagespeed-insights', true, googlesitekit )
+		).toStrictEqual(
```
Comment (felixarntz) [inline assets/js/util/test/getReAuthUrl.js]: After removing the API key, we can also remove two of these four checks.

File: assets/js/components/settings/settings-module.js [hunk 17]
```diff
@@ -39,11 +39,14 @@ import Spinner from 'GoogleComponents/spinner';
...
-const { __, sprintf } = wp.i18n;
-const { filter, map } = lodash;
-const { applyFilters } = wp.hooks;
-const { withFilters } = wp.components;
+/**
+ * WordPress dependencies
+ */
+import { Component, Fragment } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+import { filter, map } from 'lodash';
```
Comment (aaemnnosttv) [inline assets/js/components/settings/settings-module.js]: Lodash mixed in with WordPress dependencies

File: .eslintrc.json (hunk 10 above)
Comment (swissspidy) [reply .eslintrc.json]: Removing it entirely discourages the use of the `wp` global in anyway. There should be no reason to do that. And if you really need to, you can still use `const { wp } = window`

File: assets/js/components/settings/settings-modules.js [hunk 18]
```diff
@@ -28,10 +28,13 @@ import Notification from 'GoogleComponents/notifications/notification';
...
 
-const { __ } = wp.i18n;
-const { map, filter, sortBy } = lodash;
-const { Component, Fragment } = wp.element;
-const { applyFilters } = wp.hooks;
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { map, filter, sortBy } from 'lodash';
```
Comment (aaemnnosttv) [inline assets/js/components/settings/settings-modules.js]: Lodash mixed in with WordPress dependencies

File: assets/js/modules/analytics/dashboard/dashboard-alltraffic-widget-top-acquisition-sources-table.js [hunk 19]
```diff
@@ -30,9 +30,12 @@ import PreviewTable from 'GoogleComponents/preview-table';
...
 import { trafficSourcesReportDataDefaults } from '../util';
 
-const { __ } = wp.i18n;
-const { map } = lodash;
-const { Component, Fragment } = wp.element;
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { map } from 'lodash';
```
Comment (aaemnnosttv) [inline assets/js/modules/analytics/dashboard/dashboard-alltraffic-widget-top-acquisition-sources-table.js]: Lodash mixed in with WordPress dependencies

File: assets/js/util/test/__snapshots__/showErrorNotification.js.snap (hunk 12 above)
Comment (swissspidy) [reply assets/js/util/test/__snapshots__/showErrorNotification.js.snap]: Yes, that's the whole point of snapshot testing.

Learn more: https://jestjs.io/docs/en/snapshot-testing

File: stories/settings.stories.js [hunk 20]
```diff
@@ -70,7 +76,7 @@ const setupSettings = () => {
 				created: '2016-12-06T14:48:54.188Z',
 				dataRetentionResetOnNewActivity: true,
 				dataRetentionTtl: 'MONTHS_26',
-				defaultProfileId: '12345678',
+				defaultProfileID: '12345678',
```
Comment (felixarntz) [inline stories/settings.stories.js]: I think this needs to remain as is, since it is used to mock a 3P API response.

Comment (tofumatt) [reply stories/settings.stories.js]: Whoops, okay. Changing it has no effects on the tests, but if it's a mock response that makes sense 👍

Comment (aaemnnosttv) [reply stories/settings.stories.js]: @felixarntz there are other keys like this already in develop that have been changed to use our casing that shouldn't have been. I'll open a separate PR for this.
